### PR TITLE
Refactor package structure

### LINKED ISSUE #45: Improve quality of automated tests
* [ ] Separate integration tests by case, table tests
* [ ] Introduce test helpers, make tests more readable
* [ ] Add unit tests

### LINKED ISSUE #55: Cannot upload files larger than 5GB to AWS S3 in a single PUT operation
## Expected Behavior
Successfully upload files larger than 5GB (using Multipart Upload API).

## Actual Behavior
[WARNING] could not build cache, process rebuild failed, could not upload: could not upload file: couldn't put the object: EntityTooLarge: Your proposed upload exceeds the maximum allowed size

## Steps to Reproduce the Problem

  1. Build cache for a directory larger than 5GB

## Specifications

  - Image: meltwater/drone-cache:latest
  - Drone Version: 1.3.1

### LINKED ISSUE #72: Feature request: log gzip compression ratio
**Is your feature request related to a problem? Please describe.**
Currently I'm investigating a slowdown in the caching steps for a build. I'm aware the data being cached has grown significantly, so I wondered how "just tar" vs tar with gzip would perform.

Turns out, tar without gzip performed worse because I was getting a whopping 80% size reduction.

**Describe the solution you'd like**
In the archiving (and extraction?) steps, I would like to see an additional line that prints out the gzip compression ratio achieved.

**Describe alternatives you've considered**
Duplicate the caching steps, modify the copy to use tar at a different cache key, compare filesizes at the backend (in my case minio's GUI).

**Additional context**
Example build with and without gzip.
https://drone.teamopen.dev/teamopen-dev/sourcecred-stack-lookup/79

Results in the layers:
```
# With gzip
.scores 72.84 MB
.sourcecred 768.31 MB

# Without gzip
.scores 535.22 MB
.sourcecred 3.61 GB
```

### LINKED ISSUE #74: Add timeout option to cancel outgoing request after the given time
**Is your feature request related to a problem? Please describe.**
In order to prevent hanging outgoing calls to remote storages introduce a timeout logic.

**Describe the solution you'd like**
Add timeout CLI option to cancel outgoing request after the given time.
Use `context` to cancel outgoing requests.

### LINKED ISSUE #79: Add Supported storage backend section
**Is your feature request related to a problem? Please describe.**
Add a loud and clear supported backend section to the README.md and DOCS.md

### LINKED ISSUE #87: Cache cannot rebuid properly when nested paths specified in mount
**Describe the bug**
I try cache yarn.lock file, but when it's extracted from tar it md5 checksum different. This leads to reinstall node_modules dependencies.

**To Reproduce**
1. Check md5sum before cache and after extract.

**Expected behavior**
md5sum file before and after must match

**Maintainer's note:**
> Rebuild step fails to find files when nested path given:
> `packages/test/node_modules`

### LINKED ISSUE #88: When can you provide a s3 fixed-size patch upload method?
## Expected Behavior
Hi，I hope the backend of s3 can support uploading through shards

## Actual Behavior
Actually only supports PUT upload

### LINKED ISSUE #97: Customize cache key with path
The final cache path is generated by joining repo name and cache key ([here](https://github.com/meltwater/drone-cache/blob/50c33170f01195f7900e8f6457b1b8fe0dc4d53b/plugin/plugin.go#L173) & [here](https://github.com/meltwater/drone-cache/blob/50c33170f01195f7900e8f6457b1b8fe0dc4d53b/plugin/plugin.go#L148)). This prevents end users to fully control the cache destination (basically the structure is somewhat fixed: <bucket>/<repo_name>/<cache key>).

Wonder if the user could control it freely.

Regards.

### PR DESCRIPTION
**Pull request by kakkoyun**
This PR changes the package structure to make it more maintainable. This need has risen after several backends merged. 

Fixes several issues.
Improves automated tests.
Adopts stream-based IO processing.

## Proposed Changes
- Log actual compression ratio
  - Fixes #72.
  - Log actually written bytes.
  - Log final artefact size in bytes.
  - Log ratio.
- Refactor packages to make it more readable, testable and maintainable after the introduction of multiple storage backends.
  - Add missing documentation.
  - Introduce the `archive` package.
  - Introduce the `storage` package.
  - Introduce stream-based io ops.
- Update contributing docs.
- Add more markdown static change.
- AWS Multipart uploads
  - Fixes #55.
  - Fixes #88.
- Update README, add backends. 
  - Fixes #79. 
- Add timeout option.
  - Fixes #74.
- Improve tests.
  - Fixes #45. 
- Handle nested paths.
  - Fixes #87.
- Make local and remote root configurable.
  - Fixes #97. 

### Checklist

- [x] Read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] Read the [**CODE OF CONDUCT**](CODE_OF_CONDUCT.md) document.
- [x] Add tests to cover changes.
- [x] Ensure your code follows the code style of this project.
- [x] Ensure CI and all other PR checks are green OR
    - [x] Code compiles correctly.
    - [x] Created tests which fail without the change (if possible).
    - [x] All new and existing tests passed.
- [x] Add your changes to `Unreleased` section of [CHANGELOG](CHANGELOG.md).
- [x] Improve and update the [README](README.md) (if necessary).
- [x] Ensure [documentation](./DOCS.md) is up-to-date. The same file will be updated in [plugin index](https://github.com/drone/drone-plugin-index/blob/master/content/meltwater/drone-cache/index.md) when your PR is accepted, so it will be available for end-users at http://plugins.drone.io.

### CODE REVIEW

**Comment by kakkoyun** [comment]
Update for the eyes of accidental reviewer: 

This branch has gotten out of control. however please bear with me. It introduces a major overhaul on the structure and flow. This changes had been urged after the addition of several storage backends to the project. Roughly %70 of the work is done. It mostly misses tests, it'd be ready after necessary tests are added.

All the issue will that would be fixed when this got merged is listed in the description.

**Comment by aweris** [comment]
I think this PR too big to merge. I think you should split this PR to a couple of smaller PR's

File: cache/cache.go [hunk 1]
```diff
@@ -3,231 +3,81 @@ package cache
...
-	"os"
-	"path/filepath"
-	"time"
 
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
 	"github.com/meltwater/drone-cache/archive"
-	"github.com/meltwater/drone-cache/internal/metadata"
 	"github.com/meltwater/drone-cache/key"
+	"github.com/meltwater/drone-cache/key/generator"
```
Comment (aweris) [inline cache/cache.go]: I think you should restructure the last 3 commits.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -61,5 +61,7 @@ the requirements below.
     $ git push --tags
     > ...
     ```
+6. Check whether all the generate artifacts in-place properly.
```
Comment (aweris) [inline CONTRIBUTING.md]: Typo in commit message

**Comment by kakkoyun** [comment]
@aweris Thanks for the great review and feedback. After I'm done, I will separate it into more digestible chunks a thorough review.

File: Makefile [hunk 3]
```diff
@@ -10,46 +10,47 @@ GOLANGCI_LINT_VERSION=v1.21.0
 GOLANGCI_LINT_BIN=$(GOPATH)/bin/golangci-lint
 EMBEDMD_BIN=$(GOPATH)/bin/embedmd
 GOTEST_BIN=$(GOPATH)/bin/gotest
+GORELEASER_VERSION=v0.120
+GORELEASER_BIN=$(GOPATH)/bin/goreleaser
+LICHE_BIN=$(GOPATH)/bin/liche
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply Makefile:15]: I will address this in another PR

**Comment by kakkoyun** [comment]
I have cleaned the commit history. This is now ready to merge.
Thanks, @aweris and silent other for the comments.

**Comment by kakkoyun** [comment]
I'm gonna go ahead and merge this. I'll address if someone reviews it (unlikely :)) and find an issue.

**Pull request merged**